src/primitives/iter.rs

```diff
@@ -0,0 +1,521 @@
+            checksum_engine: checksum::Engine::new(),
+        }
+    }
+
+    /// Adapts a [`Checksummed`] iterator by prepending a witness version.
+    ///
+    /// Must be called before the `Checksummed` iterator has yielded any results.
+    pub fn with_witness_version(mut self, witness_version: Fe32) -> Checksummed<I, Ck> {
+        self.witness_version = Some(witness_version);
+        self
+    }
+
+    /// Adapts a [`Checksummed`] iterator into an `HrpChecksummed` iterator.
+    ///
+    /// Must be called before the `Checksummed` iterator has yielded any results.
+    pub fn hrp_checksummed(self, hrp: &Hrp) -> HrpChecksummed<I, Ck> {
+        HrpChecksummed::new(hrp, self)
+    }
+}
+
+impl<I, Ck> Iterator for Checksummed<I, Ck>
+where
+    I: Iterator<Item = Fe32>,
+    Ck: Checksum,
+{
+    type Item = Fe32;
+
+    fn next(&mut self) -> Option<Fe32> {
+        if let Some(witness_version) = self.witness_version {
+            self.witness_version = None;
+            self.checksum_engine.input_fe(witness_version);
+            return Some(witness_version);
+        }
+
+        match self.iter.next() {
+            Some(fe) => {
+                self.checksum_engine.input_fe(fe);
+                Some(fe)
+            }
+            None =>
+                if self.checksum_remaining == 0 {
+                    None
+                } else {
+                    if self.checksum_remaining == Ck::CHECKSUM_LENGTH {
+                        self.checksum_engine.input_target_residue();
+                    }
+                    self.checksum_remaining -= 1;
+                    Some(Fe32(self.checksum_engine.residue().unpack(self.checksum_remaining)))
+                },
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let wit = match self.witness_version {
+            Some(_) => 1,
+            None => 0,
+        };
+        let add = wit + self.checksum_remaining;
+        let (min, max) = self.iter.size_hint();
+
+        (min + add, max.map(|max| max + add))
+    }
+}
+
+/// Iterator adaptor for a checksummed iterator that inputs the HRP into the checksum algorithm
+/// before yielding the data followed by the checksum.
+#[derive(Clone, PartialEq, Eq)]
+pub struct HrpChecksummed<'hrp, I, Ck>
+where
+    I: Iterator<Item = Fe32>,
+    Ck: Checksum,
+{
+    /// The HRP to input into the checksum algorithm.
+    hrp: &'hrp Hrp,
+    /// A field element iterator with or without the witness version.
+    iter: Checksummed<I, Ck>,
+}
+
+impl<'hrp, I, Ck> HrpChecksummed<'hrp, I, Ck>
+where
+    I: Iterator<Item = Fe32>,
+    Ck: Checksum,
+{
+    /// Adapts an checksummed iterator by inputting the HRP to the checksum algorithm.
+    ///
+    /// `iter` may or may not include the witness version.
+    pub fn new(hrp: &'hrp Hrp, mut iter: Checksummed<I, Ck>) -> HrpChecksummed<'hrp, I, Ck> {
+        iter.checksum_engine.input_hrp(hrp);
+        Self { hrp, iter }
+    }
```

Comment:
In caa79f50d9d74b99775dfcc14f66c4140e20397c:

This `HrpChecksummed` doesn't actually output any of the HRP field elements. It needs to do that, otherwise the data you're outputting can't be validated or otherwise manipulated.

Reply:
oh, that is different to what I had in mind. I'll have to think on that a bit.